funfact/quantum/_qgates.py

```diff
@@ -0,0 +1,225 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+from abc import ABC, abstractmethod
+from numbers import Integral
+from typing import List
+from funfact import active_backend as ab
+import numpy as np
+from funfact.parametrized import Generator
+from funfact.lang._tsrex import TsrEx, tensor
+from funfact.lang._ast import Primitives as P
+from funfact.lang._terminal import ParametrizedAbstractTensor
+from funfact.lang._special import eye
+from ._special import proj0, proj1
+from funfact.conditions import Unitary
+# TODO: how to handle dtype? Should always be complex!
+
+
+class Gate(ABC):
+    '''A common base class for all gate objects.'''
+    def __init__(self, qubits: List[Integral]):
+        '''A gate stores the minimum and maximum qubits it acts on.'''
+        for i, q in enumerate(qubits):
+            if not (isinstance(q, Integral) and q >= 0):
+                raise RuntimeError(
+                    f"Qubit indices must be non-negative integers, got {q} for"
+                    f" entry {i}."
+                )
+        qubits.sort()
+        self._qubits = qubits
+
+    @abstractmethod
+    def to_tensor(self):
+        '''A gate can be converted to a tensor, either parametrized or not.'''
+        pass
+
+    @abstractmethod
+    def label(self):
+        '''A gate has a label.'''
+        pass
+
+    @property
+    def qubits(self):
```

Comment:
Is there a way that we can hint this is a list of 0-based indices, i.e. the ids of the qubits?

Reply:
Do you mean hint in the doc string?

Reply:
I've added a hint in the doc string now